src/math/interpolator.hh

```diff
@@ -0,0 +1,1168 @@
+      // One could extend this with a configurable exponential basis factor
+      // default would be 2
+    };
+
+    enum class InterpolationMethod_t { CubicSpline, CubicSplineVectorized };
+
+    template <InterpolationMethod_t Type>
+    struct InterpolationMethod {};
+
+    // An implementation of the cubic spline method, the implementation is
+    // adapted from "Numerical Recipes" [1] and optimized for uniform grids.
+    //
+    // [1] Press, William H., et al. Numerical recipes 3rd edition: The art of
+    // scientific computing. Cambridge university press, 2007.
+    template <>
+    class InterpolationMethod<InterpolationMethod_t::CubicSpline> {
+     public:
+      constexpr static InterpolationMethod_t Method{
+          InterpolationMethod_t::CubicSpline};
+
+      InterpolationMethod<InterpolationMethod_t::CubicSpline>() {}
+
+      void initialize(const Vector_Ref & grid,
+                      const Vector_Ref & evaluated_grid) {
+        this->h = grid(1) - grid(0);
+        this->h_sq_6 = this->h * this->h / 6.0;
+        this->compute_second_derivative(evaluated_grid);
+      }
+
+      void initialize(const Vector_Ref & grid,
+                      const Vector_Ref & evaluated_grid, double yd1,
+                      double ydn) {
+        this->h = grid(1) - grid(0);
+        this->h_sq_6 = this->h * this->h / 6.0;
+        this->compute_second_derivative(evaluated_grid, yd1, ydn);
+      }
+
+      inline double interpolate(const Vector_Ref & grid,
+                                const Vector_Ref & evaluated_grid, double x,
+                                int nearest_grid_index_to_x) {
+        return this->raw_interpolate(grid, evaluated_grid,
+                                     nearest_grid_index_to_x, x);
+      }
+      inline double interpolate_derivative(const Vector_Ref & grid,
+                                           const Vector_Ref & evaluated_grid,
+                                           double x,
+                                           int nearest_grid_index_to_x) {
+        return this->raw_interpolate_derivative(grid, evaluated_grid,
+                                                nearest_grid_index_to_x, x);
+      }
+
+     private:
+      // The two functions below implement the tridiagonal matrix algorithm
+      // to solve the linear system in
+      // http://mathworld.wolfram.com/CubicSpline.html
+      // under different boundary conditions. We describe with s the spline
+      // function and with y the targeted function for interpolation.
+      //
+      // Reference:
+      // https://en.wikipedia.org/wiki/Tridiagonal_matrix_algorithm
+
+      // clamped boundary conditions when first derivative for boundaries is
+      // known s'(x_0) = y'(x_0), s'(x_n) = y'(x_n)
+      inline void compute_second_derivative(const Vector_Ref & yv, double yd1,
+                                            double ydn) {
+        // Bad xa input to routine splint
+        assert(this->h != 0.0);
```

Comment:
please change float comparison to something like `abs(this->h) < TOL` where `TOL` is something really small or sufficiently small.